### ISSUE TITLE
Mapping fails when objects are single letter named

### ISSUE THREAD

**Issue by deepu105** [mapstruct/mapstruct#789]
Consider the below objects

## A.java
```
package com.mycompany.myapp.domain;

import javax.persistence.*;
import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.Objects;

/**
 * A A.
 */
public class A implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    private Long id;

    @OneToMany(mappedBy = "a")
    private Set<B> bs = new HashSet<>();

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Set<B> getBs() {
        return bs;
    }

    public void setBs(Set<B> bs) {
        this.bs = bs;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        A a = (A) o;
        if(a.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, a.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "A{" +
            "id=" + id +
            '}';
    }
}

```

## B.java

```
package com.mycompany.myapp.domain;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Objects;

/**
 * A B.
 */
public class B implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    private Long id;

    @ManyToOne
    private A a;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public A getA() {
        return a;
    }

    public void setA(A a) {
        this.a = a;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        B b = (B) o;
        if(b.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, b.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "B{" +
            "id=" + id +
            '}';
    }
}
```

## ADTO.java

```
package com.mycompany.myapp.web.rest.dto;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.Objects;


/**
 * A DTO for the A entity.
 */
public class ADTO implements Serializable {

    private Long id;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        ADTO aDTO = (ADTO) o;

        if ( ! Objects.equals(id, aDTO.id)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "ADTO{" +
            "id=" + id +
            '}';
    }
}

```

## BDTO.java

```
package com.mycompany.myapp.web.rest.dto;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.Objects;


/**
 * A DTO for the B entity.
 */
public class BDTO implements Serializable {

    private Long id;

    private Long aId;
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getAId() {
        return aId;
    }

    public void setAId(Long aId) {
        this.aId = aId;
    }
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        BDTO bDTO = (BDTO) o;

        if ( ! Objects.equals(id, bDTO.id)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "BDTO{" +
            "id=" + id +
            '}';
    }
}

```

## AMapper.java
```
package com.mycompany.myapp.web.rest.mapper;

import com.mycompany.myapp.domain.*;
import com.mycompany.myapp.web.rest.dto.ADTO;

import org.mapstruct.*;
import java.util.List;

/**
 * Mapper for the entity A and its DTO ADTO.
 */
@Mapper(componentModel = "spring", uses = {})
public interface AMapper {

    ADTO aToADTO(A a);

    List<ADTO> asToADTOs(List<A> as);

    @Mapping(target = "bs", ignore = true)
    A aDTOToA(ADTO aDTO);

    List<A> aDTOsToAS(List<ADTO> aDTOs);
}

```

## BMapper.java
```
package com.mycompany.myapp.web.rest.mapper;

import com.mycompany.myapp.domain.*;
import com.mycompany.myapp.web.rest.dto.BDTO;

import org.mapstruct.*;
import java.util.List;

/**
 * Mapper for the entity B and its DTO BDTO.
 */
@Mapper(componentModel = "spring", uses = {})
public interface BMapper {

    @Mapping(source = "a.id", target = "aId")
    BDTO bToBDTO(B b);

    List<BDTO> bsToBDTOs(List<B> bs);

    @Mapping(source = "aId", target = "a")
    B bDTOToB(BDTO bDTO);

    List<B> bDTOsToBS(List<BDTO> bDTOs);

    default A aFromId(Long id) {
        if (id == null) {
            return null;
        }
        A a = new A();
        a.setId(id);
        return a;
    }
}
```

## Error

This throws below error during compile
```
[ERROR] diagnostic: D:\Projects\test\testApp\src\main\java\com\mycompany\myapp\web\rest\mapper\BMapper.java:20: error: No property named "aId" exists in source parameter(s).
    @Mapping(source = "aId", target = "a")
    ^
[ERROR] diagnostic: D:\Projects\test\testApp\src\main\java\com\mycompany\myapp\web\rest\mapper\BMapper.java:15: error: Unknown property "aId" in return type.
    @Mapping(source = "a.id", target = "aId")
    ^
[ERROR] error on execute: error during compilation
```

**Comment by agudian** [comment]
Well, I took a closer look why that happens. The thing is, that according to the Java Bean Specification the correct accessor methods for the property `aId` would have to be `getaId()` and `setaId(..)`, as the property name derived from the accessors `getAId()` / `setAId(..)` would be `AId`.

(from the bean specification 1.01, section 8.8):
> Thus when we extract a property or event name from the middle of an existing Java name, we
normally convert the first character to lower case. However to support the occasional use of all
upper-case names, we check if the first two characters of the name are both upper case and if
so leave it alone.

MapStruct works on property names derived from the accessor methods and not the actual class attribute names. In your case, you'd have to use `"AId"` in the `@Mapping` annotations.

This may look funny. But I think we should stick to the bean-spec here. Otherwise, what would be the harm being a bit more lax in case we don't find a match with the strict checking... 
I guess we should think about that...

@gunnarmorling / @sjaakd: guys, what's your take on that?

**Comment by deepu105** [comment]
Oh shit the the funny bean spec. Agree we should stick to that even if its
funny. I guess we will make changes then. You can close this then

**Comment by sjaakd** [comment]
I encountered similar problems with code generation in XJC. It uses XML definitions, and some weeks ago I came across a similar issue together with a colleague of mine.

However, there is also an ongoing discussion at my work. I've encountered developers that always want to specify all mappings and others (including myself) that want to leave as much up to MapStruct. Belonging to the latter fraction :smile: 

Being of the latter fraction, I still would like to plea for a regular based expression mechanism on configuration level for name matching. That would solve this issue. But also include DB normalization that I discussed before (like booleans are by our DB architects always have the extension BLN whether the interface - XSD - designers omit it). Combining this with good MapStruct loggin on decision making would help.
